packages/teleport/src/lib/tdp/codec.ts

```diff
@@ -328,26 +343,34 @@ export default class Codec {
     return message;
   }
 
-  // decodeRegion decodes the region from a PNG_FRAME tdp message.
-  decodeRegion(buffer: ArrayBuffer): Region {
+  // decodePngFrame decodes a raw tdp PNG frame message and returns it as a PngFrame
+  // | message type (2) | left uint32 | top uint32 | right uint32 | bottom uint32 | data []byte |
+  // https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#2---png-frame
+  decodePngFrame(buffer: ArrayBuffer, onload: (PngFrame) => any): PngFrame {
     let dv = new DataView(buffer);
-    return {
+    const image = new Image();
+    const pngFrame = {
       left: dv.getUint32(1),
       top: dv.getUint32(5),
       right: dv.getUint32(9),
       bottom: dv.getUint32(13),
+      data: image,
     };
+    pngFrame.data.onload = onload(pngFrame);
+    pngFrame.data.src = this._decodePng(buffer);
+
+    return pngFrame;
   }
 
   // Taken as the winning algorithm of https://jsbench.me/vjk9nczxst/1
   // jsbench link was discovered in https://gist.github.com/jonleighton/958841
-  toBase64(buffer: ArrayBuffer) {
+  _toBase64(buffer: ArrayBuffer) {
     const binary = String.fromCharCode.apply(null, new Uint8Array(buffer, 17));
     return btoa(binary);
   }
 
   // decodePng creates a data:image uri from the png data part of a PNG_FRAME tdp message.
-  decodePng(buffer: ArrayBuffer): string {
-    return `data:image/png;base64,${this.toBase64(buffer)}`;
+  _decodePng(buffer: ArrayBuffer): string {
```

Comment:
Is there a better name for this? You're not actually decoding PNG data. If anything, you're _encoding_ base64 data.